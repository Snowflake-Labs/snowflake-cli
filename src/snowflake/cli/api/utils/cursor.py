from __future__ import annotations

from typing import Callable, Iterable, List, Optional

from snowflake.connector.cursor import DictCursor, SnowflakeCursor


def _rows_generator(cursor: DictCursor, predicate: Callable[[dict], bool]):
    return (row for row in cursor.fetchall() if predicate(row))


def find_all_rows(cursor: DictCursor, predicate: Callable[[dict], bool]) -> List[dict]:
    return list(_rows_generator(cursor, predicate))


def find_first_row(
    cursor: DictCursor, predicate: Callable[[dict], bool]
) -> Optional[dict]:
    """Returns the first row that matches the predicate, or None."""
    return next(_rows_generator(cursor, predicate), None)


def join_cursors(cursors: List[Iterable[SnowflakeCursor]]) -> Iterable[SnowflakeCursor]:
    while cursors:
        cur = cursors.pop(0)
        yield from cur
