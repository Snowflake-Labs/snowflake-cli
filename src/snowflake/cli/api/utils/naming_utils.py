from __future__ import annotations

import re
from typing import Tuple

from snowflake.cli.api.project.util import (
    VALID_IDENTIFIER_REGEX,
)


def from_qualified_name(name: str) -> Tuple[str, str | None, str | None]:
    """
    Takes in an object name in the form [[database.]schema.]name. Returns a tuple (name, [schema], [database])
    """
    # TODO: Use regex to match object name to a valid identifier or valid identifier (args). Second case is for sprocs and UDFs
    qualifier_pattern = rf"(?:(?P<first_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?:(?P<second_qualifier>{VALID_IDENTIFIER_REGEX})\.)?(?P<name>.*)"
    result = re.fullmatch(qualifier_pattern, name)

    if result is None:
        raise ValueError(f"'{name}' is not a valid qualified name")

    unqualified_name = result.group("name")
    if result.group("second_qualifier") is not None:
        database = result.group("first_qualifier")
        schema = result.group("second_qualifier")
    else:
        database = None
        schema = result.group("first_qualifier")
    return unqualified_name, schema, database
