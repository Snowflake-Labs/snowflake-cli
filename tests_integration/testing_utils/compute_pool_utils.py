import math
import time

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_executed_successfully,
)


class ComputePoolTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session


class ComputePoolTestSteps:
    def __init__(self, setup: ComputePoolTestSetup):
        self._setup = setup

    def create_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "compute-pool",
                "create",
                compute_pool_name,
                "--min-nodes",
                1,
                "--family",
                "CPU_X64_XS",
            ]
        )
        assert result.json, result.output
        assert "status" in result.json
        assert (
            f"Compute Pool {compute_pool_name.upper()} successfully created."
            in result.json["status"]  # type: ignore
        )

    def list_should_return_compute_pool(self, compute_pool_name) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def list_should_not_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": compute_pool_name.upper()})

    def describe_should_return_compute_pool(self, compute_pool_name: str) -> None:
        result = self._execute_describe(compute_pool_name)
        assert result.json
        assert len(result.json) == 1
        assert result.json[0]["name"] == compute_pool_name.upper()

    def drop_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "compute-pool",
                compute_pool_name,
            ],
        )
        assert result.json
        assert len(result.json) == 1
        assert result.json[0] == {  # type: ignore
            "status": f"{compute_pool_name.upper()} successfully dropped."
        }

    def stop_all_on_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "stop-all", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def suspend_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "suspend", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def resume_compute_pool(self, compute_pool_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            ["spcs", "compute-pool", "resume", compute_pool_name]
        )
        assert_that_result_is_successful_and_executed_successfully(result, is_json=True)

    def wait_until_compute_pool_is_idle(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "IDLE", 300)

    def wait_until_compute_pool_is_suspended(self, compute_pool_name: str) -> None:
        self._wait_until_compute_pool_reaches_state(compute_pool_name, "SUSPENDED", 60)

    def _wait_until_compute_pool_reaches_state(
        self, compute_pool_name: str, target_state: str, max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        target_state = target_state.upper()
        for i in range(max_counter):
            status = self._execute_describe(compute_pool_name)
            if contains_row_with(status.json, {"state": target_state}):
                return
            time.sleep(10)
        status = self._execute_describe(compute_pool_name)

        error_message = f"Compute pool {compute_pool_name} didn't reach target state '{target_state}' in {max_duration} seconds. Current state is '{status.json['state']}'"
        pytest.fail(error_message)

    def _execute_describe(self, compute_pool_name: str):
        return self._setup.runner.invoke_with_connection_json(
            ["object", "describe", "compute-pool", compute_pool_name]
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "compute-pool",
            ],
        )
