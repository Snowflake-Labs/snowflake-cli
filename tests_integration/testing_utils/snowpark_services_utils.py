import json
import math
import time
from textwrap import dedent
from typing import Union

import pytest
from snowflake.connector import SnowflakeConnection

from tests_integration.conftest import SnowCLIRunner
from tests_integration.test_utils import contains_row_with, not_contains_row_with
from tests_integration.testing_utils.assertions.test_result_assertions import (
    assert_that_result_is_successful_and_output_json_contains,
    assert_that_result_is_successful_and_output_json_equals,
)


class SnowparkServicesTestSetup:
    def __init__(
        self,
        runner: SnowCLIRunner,
        snowflake_session: SnowflakeConnection,
        test_root_path,
    ):
        self.runner = runner
        self.snowflake_session = snowflake_session
        self.test_root_path = test_root_path


class SnowparkServicesTestSteps:
    compute_pool = "snowcli_compute_pool"
    database = "snowcli_db"
    schema = "public"
    container_name = "echo-test"

    def __init__(self, setup: SnowparkServicesTestSetup):
        self._setup = setup

    def create_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "create",
                service_name,
                "--compute-pool",
                self.compute_pool,
                "--spec-path",
                self._get_spec_path(),
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert_that_result_is_successful_and_output_json_equals(
            result, {"status": f"Service {service_name.upper()} successfully created."}
        )

    def status_should_return_service(self, service_name: str) -> None:
        result = self._execute_status(service_name)
        assert_that_result_is_successful_and_output_json_contains(
            result,
            {"containerName": self.container_name, "serviceName": service_name.upper()},
        )

    def logs_should_return_service_logs(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection(
            [
                "spcs",
                "service",
                "logs",
                service_name,
                "--container-name",
                self.container_name,
                "--instance-id",
                "0",
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )
        assert result.output
        # Assert this instead of full payload due to log coloring
        assert service_name in result.output
        assert '"GET /healthcheck HTTP/1.1" 200 -' in result.output

    def list_should_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert contains_row_with(result.json, {"name": service_name.upper()})

    def list_should_not_return_service(self, service_name: str) -> None:
        result = self._execute_list()
        assert not_contains_row_with(result.json, {"name": service_name.upper()})

    def describe_should_return_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "describe",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json
        assert result.json[0]["name"] == service_name.upper()  # type: ignore

    def drop_service(self, service_name: str) -> None:
        result = self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "drop",
                "service",
                f"{self.database}.{self.schema}.{service_name}",
            ],
        )
        assert result.json[0] == {  # type: ignore
            "status": f"{service_name.upper()} successfully dropped."
        }

    def wait_until_service_is_ready(self, service_name: str) -> None:
        self._wait_until_service_reaches_state(service_name, "READY", 900)

    def _wait_until_service_reaches_state(
        self, service_name: str, target_status: Union[str, dict], max_duration: int
    ):
        assert max_duration > 0
        max_counter = math.ceil(max_duration / 10)
        if isinstance(target_status, str):
            target_status = {"status": target_status}
        for i in range(max_counter):
            status = self._execute_status(service_name)
            if contains_row_with(status.json, target_status):
                return
            elif contains_row_with(status.json, {"status": "FAILED"}):
                describe = self._setup.runner.invoke_with_connection_json(
                    ["object", "describe", "service", service_name]
                )
                pytest.fail(
                    dedent(
                        f"""
                    {service_name} service failed before reaching target state:
                    {json.dumps(target_status)}
                    current state:
                    {json.dumps(status)}
                    current describe:
                    {json.dumps(describe)}
                    """
                    )
                )
            time.sleep(10)
        status = self._execute_status(service_name)

        error_message = f"""
{service_name} service didn't reach target state in {max_duration} seconds.
target:
{json.dumps(target_status)}
current:
{json.dumps(status.json)}
"""
        pytest.fail(error_message)

    def _execute_status(self, service_name: str):
        return self._setup.runner.invoke_with_connection_json(
            [
                "spcs",
                "service",
                "status",
                service_name,
                "--database",
                self.database,
                "--schema",
                self.schema,
            ],
        )

    def _execute_list(self):
        return self._setup.runner.invoke_with_connection_json(
            [
                "object",
                "list",
                "service",
            ],
        )

    def _get_spec_path(self):
        return f"{self._setup.test_root_path}/spcs/spec/spec.yml"
