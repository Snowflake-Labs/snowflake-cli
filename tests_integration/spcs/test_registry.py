import pytest
from snowflake.connector.cursor import DictCursor

from tests_integration.testing_utils.naming_utils import ObjectNameProvider


@pytest.mark.integration
def test_token(runner):
    result = runner.invoke_with_connection_json(["spcs", "image-registry", "token"])

    assert result.exit_code == 0
    assert result.json
    assert "token" in result.json
    assert result.json["token"]
    assert "expires_in" in result.json
    assert result.json["expires_in"]


@pytest.mark.integration
def test_get_registry_url(test_database, test_role, runner, snowflake_session):
    # newly created role should have no access to image repositories and should not be able to get registry URL
    test_repo = ObjectNameProvider("test_repo").create_and_get_next_object_name()
    snowflake_session.execute_string(f"create image repository {test_repo}")

    fail_result = runner.invoke_with_connection(
        ["spcs", "image-registry", "url", "--role", test_role]
    )
    assert fail_result.exit_code == 1, fail_result.output
    assert "Current role cannot view any image repositories." in fail_result.output

    # role should be able to get registry URL once granted read access to an image repository
    repo_list_cursor = snowflake_session.execute_string(
        "show image repositories", cursor_class=DictCursor
    )
    expected_repo_url = repo_list_cursor[0].fetchone()["repository_url"]
    expected_registry_url = "/".join(expected_repo_url.split("/")[:-3])
    snowflake_session.execute_string(
        f"grant usage on database {snowflake_session.database} to role {test_role};"
        f"grant usage on schema {snowflake_session.schema} to role {test_role};"
        f"grant read on image repository {test_repo} to role {test_role};"
    )
    success_result = runner.invoke_with_connection(
        ["spcs", "image-registry", "url", "--role", test_role]
    )
    assert success_result.exit_code == 0, success_result.output
    assert success_result.output.strip() == expected_registry_url
