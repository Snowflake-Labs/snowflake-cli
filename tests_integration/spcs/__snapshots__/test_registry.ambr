# serializer version: 1
# name: test_get_registry_url
  '''
  ╭─ Error ──────────────────────────────────────────────────────────────────────╮
  │ Current role cannot view any image repositories. Please ensure that at least │
  │ one repository exists in your image registry and use a role with read access │
  │ to at least one image repository before retrieving registry URL.             │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  '''
# ---
